File: src/_cpp/Combinations.cpp
#include "Combinations.h"

NAMESPACE

KeyCombination::KeyCombination(int k1, int k2, func c) {
	key1 = k1;
	key2 = k2;
	callback = c;
}

KeyCombination::KeyCombination(int k1, int k2, int k3, func c) {
	key1 = k1;
	key2 = k2;
	key3 = k3;
	callback = c;
}

void KeyCombination::operator()() {
	callback();
}

bool KeyCombination::operator==(KeyCombination combo) {
	return key1 == combo.key1 && key2 == combo.key2 && key3 == combo.key3;
}

MouseKeyCombination::MouseKeyCombination(int k, mousecall c) {
	key = k;
	callback = c;
}

MouseKeyCombination::MouseKeyCombination(int k, func c) {
	key = k;
	callback = [c](int x, int y) {
		c();
	};
}

void MouseKeyCombination::operator()(int x, int y) {
	callback(x, y);
}

bool MouseKeyCombination::operator==(MouseKeyCombination other) {
	return key == other.key;
}

END

File: src/events/_combinations.py
class Combination:
    def __eq__(self, other):
        if self.size == other.size: #a KeyError will occur with uneven sizes
            for key in [attr for attr in dir(self) if attr[:3] == "key"]:
                if not eval(f"self.{key} == other.{key}"):
                    return False
            return True
        return False

class KeyCombination(Combination):
    size = 2
    def __init__(self, key1, key2, key3, callback):
        self.key1 = key1
        self.key2 = key2
        if key3:
            self.key3 = key3
            self.size += 1
        self.callback = callback

class MouseKeyCombination(Combination):
    size = 1
    def __init__ (self, key1, key2, key3, callback):
        self.key1 = key1
        if key2:
            self.key2 = key2
            self.size += 1
        if key3:
            self.key3 = key3
            self.size += 1
        self.callback = callback
